### PR TITLE
added various transation guard API to storage_base

### PR DESCRIPTION
**Pull request by fnc12**

### CODE REVIEW

File: tests/transaction_tests.cpp [hunk 1]
```diff
@@ -102,20 +102,60 @@ TEST_CASE("Transaction guard") {
 
     SECTION("insert, call make a storage to call an exception and check that rollback was fired") {
         auto countBefore = storage.count<Object>();
-        try {
-            auto guard = storage.transaction_guard();
+        SECTION("transaction_guard") {
+            try {
```
Comment (trueqbit) [inline tests/transaction_tests.cpp:106]: Can we use a dedicated Catch2 test macro for this, please?
```c++
REQUIRE_THROWS_WITH(storage.transaction([&] {
    storage.insert(Object{0, "John"});
    storage.get<Object>(0);
    return true;
}),
                    "Not found");
```

File: tests/transaction_tests.cpp [hunk 2]
```diff
@@ -102,20 +102,60 @@ TEST_CASE("Transaction guard") {
...
+                auto guard = storage.transaction_guard();
 
-            storage.insert(Object{0, "John"});
+                storage.insert(Object{0, "John"});
 
-            storage.get<Object>(-1);
+                storage.get<Object>(-1);
```
Comment (trueqbit) [inline tests/transaction_tests.cpp:111]: Also, the object ID to retrieve should be _**0**_, no?

**Review by trueqbit — CHANGES_REQUESTED** [review]

File: tests/transaction_tests.cpp (hunk 2 above)
Comment (fnc12) [reply tests/transaction_tests.cpp:111]: can be 0, but any value below 0 is also suitable for non-existing object

Comment (fnc12) [reply tests/transaction_tests.cpp:111]: but I confess this code was written by me during epoch when I was sure that `0` is a valid ID in SQLite

File: tests/transaction_tests.cpp (hunk 1 above)
Comment (fnc12) [reply tests/transaction_tests.cpp:106]: let me try

Comment (fnc12) [reply tests/transaction_tests.cpp:106]: actually I'd like to test exception's error code more than its description string. But I don't see any option of doing it with `REQUIRE_THROWS_WITH` so I'd like to keep the things as they are

Comment (fnc12) [reply tests/transaction_tests.cpp]: done. Please check

File: tests/transaction_tests.cpp [hunk 3]
```diff
@@ -100,22 +100,71 @@ TEST_CASE("Transaction guard") {
 
     storage.insert(Object{0, "Jack"});
 
-    SECTION("insert, call make a storage to call an exception and check that rollback was fired") {
-        auto countBefore = storage.count<Object>();
-        try {
-            auto guard = storage.transaction_guard();
+    struct ExceptionMatcher : Catch::Matchers::MatcherGenericBase {
+        std::string what;
```
Comment (trueqbit) [inline tests/transaction_tests.cpp:104]: Checking the error code alone would be enough, we don't care about the exact message string at this point...

File: tests/transaction_tests.cpp [hunk 4]
```diff
@@ -100,22 +100,71 @@ TEST_CASE("Transaction guard") {
...
+            std::stringstream ss;
+            const std::string message = get_orm_error_category().message(this->errorCode.value());
+            ss << "Exception with code: " << message << ", what: " << this->what;
+            return ss.str();
+        }
+    };
 
-            REQUIRE(countBefore == countNow);
+    const ExceptionMatcher notFoundExceptionMatcher("Not found", make_error_code(orm_error_code::not_found));
+    SECTION("insert, call make a storage to call an exception and check that rollback was fired") {
```
Comment (trueqbit) [inline tests/transaction_tests.cpp:124]: BTW, this section within this test case replaces the same section in test case "transaction_rollback" some lines above, so the old one can be removed AFAIK?

**Review by trueqbit — APPROVED** [review]

File: tests/transaction_tests.cpp (hunk 3 above)
Comment (fnc12) [reply tests/transaction_tests.cpp]: fixed

File: tests/transaction_tests.cpp (hunk 4 above)
Comment (fnc12) [reply tests/transaction_tests.cpp:122]: agree, I removed `TEST_CASE("transaction_rollback")`

**Review by trueqbit — APPROVED** [review]

**Pull request merged**